File: kotlin-client/src/main/kotlin/io/provenance/bilateral/models/AdminMatchOptions.kt
package io.provenance.bilateral.models

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeName
import com.fasterxml.jackson.databind.PropertyNamingStrategies.SnakeCaseStrategy
import com.fasterxml.jackson.databind.annotation.JsonNaming

@JsonNaming(SnakeCaseStrategy::class)
@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
sealed interface AdminMatchOptions {
    /**
     * Admin options available when executing a match for two coin trades.  Requests including coin trade admin options
     * for non-coin-trade matches will be rejected.
     *
     * @param acceptMismatchedBids If true, a match will be executed even if the bid offers a lower (or even completely
     * different denom) coin than was requested in the ask's quote.  Ex: Asker requests 200nhash and bidder offers
     * 100nhash - refused unless this flag is 'true'.  Ex: Asker requests 200nhash and bidder offers 500000dogecoin -
     * refused unless this flag is 'true'.
     */
    @JsonNaming(SnakeCaseStrategy::class)
    @JsonTypeName("coin_trade")
    data class CoinTradeAdminOptions(
        val acceptMismatchedBids: Boolean? = null,
    ) : AdminMatchOptions {
        override fun toString(): String = "coin_trade: accept_mismatched_bids = $acceptMismatchedBids"
    }

    /**
     * Admin options available when executing a match for two marker trades.  Requests including marker trade admin
     * options for non-marker-trade matches will be rejected.
     *
     * @param acceptMismatchedBids If true, a match will be executed even if the bid offers a lower (or even completely
     * different denom) coin than was requested in the ask's quote.  Ex: Asker requests 200nhash and bidder offers
     * 100nhash - refused unless this flag is 'true'.  Ex: Asker requests 200nhash and bidder offers 500000dogecoin -
     * refused unless this flag is 'true'.
     */
    @JsonNaming(SnakeCaseStrategy::class)
    @JsonTypeName("marker_trade")
    data class MarkerTradeAdminOptions(
        val acceptMismatchedBids: Boolean? = null,
    ) : AdminMatchOptions {
        override fun toString(): String = "marker_trade: accept_mismatched_bids = $acceptMismatchedBids"
    }

    /**
     * Admin options available when executing a match for two marker share sales.  Requests including marker share sale
     * admin options for non-marker-share-sale matches will be rejected.
     *
     * @param overrideQuoteSource If provided, the target's quote will be used to determine how much quote funds are
     * sent to the asker after the trade completes.  The contract will automatically calculate a quote per share value
     * for the ask or bid based on values stored with the ask or bid order, and use that value alongside the bid's
     * amount of shares being purchased to determine how much funds will be sent to the asker.
     */
    @JsonNaming(SnakeCaseStrategy::class)
    @JsonTypeName("marker_share_sale")
    data class MarkerShareSaleAdminOptions(
        val overrideQuoteSource: OverrideQuoteSource? = null,
    ) : AdminMatchOptions {
        override fun toString(): String = "marker_share_sale: override_quote_source = $overrideQuoteSource"
    }

    /**
     * Admin options available when executing a match for two scope trades.  Requests including scope trade admin
     * options for non-scope-trade matches will be rejected.
     *
     * @param acceptMismatchedBids If true, a match will be executed even if the bid offers a lower (or even completely
     * different denom) coin than was requested in the ask's quote.  Ex: Asker requests 200nhash and bidder offers
     * 100nhash - refused unless this flag is 'true'.  Ex: Asker requests 200nhash and bidder offers 500000dogecoin -
     * refused unless this flag is 'true'.
     */
    @JsonNaming(SnakeCaseStrategy::class)
    @JsonTypeName("scope_trade")
    data class ScopeTradeAdminOptions(
        val acceptMismatchedBids: Boolean? = null,
    ) : AdminMatchOptions {
        override fun toString(): String = "scope_trade: accept_mismatched_bids = $acceptMismatchedBids"
    }
}

enum class OverrideQuoteSource {
    @JsonProperty("ask")
    ASK,
    @JsonProperty("bid")
    BID,
}

File: smart-contract/src/types/request/admin_match_options.rs
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// These options are to be used in matching by the contract admin.  Requests not made by the admin
/// that include admin match options will be rejected.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum AdminMatchOptions {
    CoinTrade {
        /// Allows trades with an ask quote that does not match the bid quote to still be executed.
        /// Defaults to false if not specified.
        accept_mismatched_bids: Option<bool>,
    },
    MarkerTrade {
        /// Allows trades with an ask quote that does not match the bid quote to still be executed.
        /// Defaults to false if not specified.
        accept_mismatched_bids: Option<bool>,
    },
    MarkerShareSale {
        /// Allows trades to use either the ask quote or the bid quote to determine the funds allocated
        /// when an ask or bid is executed.  Will never try to use more than the bid quote amount
        /// due to the bid quote being the source of the available funds.
        override_quote_source: Option<OverrideQuoteSource>,
    },
    ScopeTrade {
        /// Allows trades with an ask quote that does not match the bid quote to still be executed.
        /// Defaults to false if not specified.
        accept_mismatched_bids: Option<bool>,
    },
}
#[cfg(test)]
impl AdminMatchOptions {
    pub fn coin_trade_empty() -> Self {
        Self::CoinTrade {
            accept_mismatched_bids: None,
        }
    }

    pub fn coin_trade_options(accept_mismatched_bids: bool) -> Self {
        Self::CoinTrade {
            accept_mismatched_bids: Some(accept_mismatched_bids),
        }
    }

    pub fn marker_trade_empty() -> Self {
        Self::MarkerTrade {
            accept_mismatched_bids: None,
        }
    }

    pub fn marker_trade_options(accept_mismatched_bids: bool) -> Self {
        Self::MarkerTrade {
            accept_mismatched_bids: Some(accept_mismatched_bids),
        }
    }

    pub fn marker_share_sale_empty() -> Self {
        Self::MarkerShareSale {
            override_quote_source: None,
        }
    }

    pub fn marker_share_sale_options(override_quote_source: OverrideQuoteSource) -> Self {
        Self::MarkerShareSale {
            override_quote_source: Some(override_quote_source),
        }
    }

    pub fn scope_trade_empty() -> Self {
        Self::ScopeTrade {
            accept_mismatched_bids: None,
        }
    }

    pub fn scope_trade_options(accept_mismatched_bids: bool) -> Self {
        Self::ScopeTrade {
            accept_mismatched_bids: Some(accept_mismatched_bids),
        }
    }
}

/// Determines the source of the quote to be sent to the asker after a match executes.  By standard,
/// the ask and bid should have matching quotes, but this allows them to not match and for a
/// mismatched ask or bid to drive the amount that is sent, instead.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum OverrideQuoteSource {
    Ask,
    Bid,
}
